systems/billing/collector/internal/server/event.go

```diff
@@ -0,0 +1,280 @@
+	simpb "github.com/ukama/ukama/systems/subscriber/sim-manager/pb/gen"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/anypb"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// TODO: We need to think about retry policies for failing interaction between our backend and the upstream billing service
+// provider
+
+const (
+	handlerTimeoutFactor = 3
+)
+
+type BillingCollectorEventServer struct {
+	client client.BillingClient
+	epb.UnimplementedEventNotificationServiceServer
+}
+
+func NewBillingCollectorEventServer(client client.BillingClient) *BillingCollectorEventServer {
+	return &BillingCollectorEventServer{
+		client: client,
+	}
+}
+
+func (b *BillingCollectorEventServer) EventNotification(ctx context.Context, e *epb.Event) (*epb.EventResponse, error) {
+	log.Infof("Received a message with Routing key %s and Message %+v", e.RoutingKey, e.Msg)
+
+	switch e.RoutingKey {
+
+	// Send usage event
+	case "event.cloud.cdr.sim.usage":
+		msg, err := unmarshalOperatorSimUsage(e.Msg)
+		if err != nil {
+			return nil, err
+		}
+
+		err = handleCdrSimUsageEvent(e.RoutingKey, msg, b)
+		if err != nil {
+			return nil, err
+		}
+
+	// Create customer
+	case "event.cloud.registry.subscriber.create":
+		msg, err := unmarshalSubscriber(e.Msg)
+		if err != nil {
+			return nil, err
+		}
+
+		err = handleRegistrySubscriberCreateEvent(e.RoutingKey, msg, b)
+		if err != nil {
+			return nil, err
+		}
+
+	// Update customer
+	case "event.cloud.registry.subscriber.update":
+		msg, err := unmarshalSubscriber(e.Msg)
+		if err != nil {
+			return nil, err
+		}
+
+		err = handleRegistrySubscriberUpdateEvent(e.RoutingKey, msg, b)
+		if err != nil {
+			return nil, err
+		}
+
+	// Delete customer
+	case "event.cloud.registry.subscriber.delete":
+		msg, err := unmarshalSubscriber(e.Msg)
+		if err != nil {
+			return nil, err
+		}
+
+		err = handleRegistrySubscriberDeleteEvent(e.RoutingKey, msg, b)
+		if err != nil {
+			return nil, err
+		}
+
+	// add or update subscrition to customer
+	case "event.cloud.simmanager.package.activate":
+		msg, err := unmarshalSim(e.Msg)
+		if err != nil {
+			return nil, err
+		}
+
+		err = handleSimManagerSetActivePackageForSimEvent(e.RoutingKey, msg, b)
+		if err != nil {
+			return nil, err
+		}
+
+	default:
+		log.Errorf("No handler routing key %s", e.RoutingKey)
+	}
+
+	return &epb.EventResponse{}, nil
+}
+
+func handleCdrSimUsageEvent(key string, simUsage *operatorpb.SimUsage, b *BillingCollectorEventServer) error {
+	log.Infof("Keys %s and Proto is: %+v", key, simUsage)
+
+	ctx, cancel := context.WithTimeout(context.Background(), handlerTimeoutFactor*time.Second)
+	defer cancel()
+
+	event := client.Event{
+		//TODO: To be replaced by msgClient msgId
+		TransactionId: fmt.Sprintf("%s%d", simUsage.Id, time.Now().Unix()),
+
+		CustomerId:     simUsage.SubscriberId,
+		SubscriptionId: simUsage.SimId,
+		Code:           "data_usage",
+		SentAt:         time.Now(),
```

Comment:
What is this time used for ? Hopefully not in the user invoice.

Reply:
> What is this time used for ? Hopefully not in the user invoice.

This is when the usage event is sent, and I am not sure but I think it might be used for invoice because invoice is calculated  for events that were sent within the billable period. Will  definitely have to check.

Reply:
> > What is this time used for ? Hopefully not in the user invoice.
> 
> This is when the usage event is sent, and I am not sure but I think it might be used for invoice because invoice is calculated for events that were sent within the billable period. Will definitely have to check.

Checking or not, either way  now I think this should be mapped to the ending time of the data session.  We'll update in the next iteration.